atomic_reactor/plugins/post_compare_components.py

```diff
@@ -106,14 +126,19 @@ def run(self):
                     master_comp[identifier] = component
                     continue
 
-                try:
+                if t == T_RPM:
                     mc = master_comp[identifier]
-
-                    if t == 'rpm':
+                    try:
                         self.rpm_compare(mc, component)
-                except ValueError as ex:
-                    self.log.warn("Comparison mismatch for component %s: %s", name, ex)
-                    failed = True
+                    except ValueError as ex:
+                        self.log.debug("Mismatch details: %s", ex)
+                        self.log.warning(
+                            "Comparison mismatch for component %s:", name)
+
+                        # use all components to provide complete list
+                        for comp in filter_components_by_name(name, comp_list):
```

Comment:
I think it's a bit inefficient to loop through the list again. Can we just log `mc` and `component`?

The warning will be repeated if more than two arches are used regardless.

Reply:
Alternatively, you could maintain a list of all the failures and only print warning once at the end. You could also use the component names when raising "Failed component comparison" exception to make things clearer to users, since that's what they'll see as the failure. Everything else is buried in logs.

```python
mismatches = defaultdict(list)

# rpm_compare return bool instead of raising exception
if self.rpm_compare(mc, component):
  dict[identifier].append(component)

...

# Report warnings here once
for component_name, components in mismatches.items():
  ...

if mismatches:
  raise ValueError('Failed component comparison for components: {}'.format(mismatches.keys()))

```

Reply:
How often does this happen? I'd not do premature optimization, I don't think that searching a few lists and dicts may be somehow serious.

I wanted (and got agreement in ticket) to provide complete list of all architectures. This may not happen when I compare only `mc` and `component` in order: `good1, good2, bad` and it is also less user friendly.

I will incorporate your idea of reporting only once into patches. Great idea! Shame on me I didin't do that. But it will be different because as I mentioned previously, I cannot rely only on comparison `mc` and `component`.

Reply:
BTW: I cannot use sets, dict is mutable. (May work in py2, but it is still bad in way how is that implemented)

Reply:
Slight misunderstanding. The idea is that mismatches would look something like this:
```python
mismatches = {
  (rpm, libquadmath): [
    {'name': 'libquadmath', 'version': '1.2', 'arch': 'x86_64', ...},
    {'name': 'libquadmath', 'version': '1.2', 'arch': 'ppc64le', ...},
    {'name': 'libquadmath', 'version': '1.2', 'arch': 's390x', ...},
  ]
}
```
In other words, it would include *all* the mismatches. Including `mc` and `component`. The comment about `mc` in snippet is that you may want to use something like this:
```python
mismatches = defaultdict(set)
...
if self.rpm_compare(mc, component):
  mismatches[identifier].add(component)
  mismatches[identifier].add(mc)
```
Using a set would ensure that `mc` doesn't end up in there twice :)

EDIT: Since dict can be added to a set, you could always turn it into a string since that's all you need at this point anyways.

Just an idea. Feel free to ignore it :)

Reply:
I got your point, but it won't work reliably all the time (or I'm wrong, correct me)

Let's assume that we have following packages, checked in order:
```
(1) x86_64: libquadmath-1.2-abc (987654321)
(2) ppc64le: libquadmath-1.2-abc (987654321)
(3) s390x: libquadmath-1.3.1415-abc (987654321)  # mismatch
```

In the current implementation, first entry is permanently stored as `mc`
```
store mc = (1)
compare (mc) vs (2) --> ok
compare (mc) vs (3) --> mismatch
```
so (2) will be never added into `mismatches` dict. There is no comparison `(2) vs (3)`. This will lead into incomplete list of components in the log output.

I can do mapping `{name: [all_components], ...}` but creating/updating dicts and lists is less efficient in memory and speed, than simple iteration over yielded values (in case that mismatches are for few entries only and rare).